### PR TITLE
example implementation for learning curve generation

### PR DESCRIPTION
**Pull request by vonHartz**
_Hint_: This pull request is designed as a feature request paired with an example implementation, not necessarily to directly pull it.

It would be nice if ccobra could automatically generate learning curves for the models so that one can get a sense for the learning process.

I have made an example implementation of such feature by creating model checkpoints every training epoch during both pre-training and adaption. 
If the corresponding argument is provided to the runner, all the created model checkpoints will be evaluated and learning curves are generated at the end of the run.
In my implementation the plots are made using seaborn and directly saved to the local drive.
Of course, it would be more in line with the general design to display the plots as part of the HTML-report.

### CODE REVIEW

**Comment by nriesterer** [comment]
Thanks you for your contribution!

Could you elaborate a bit on the following points:

1. What is the general goal of your extension (presumably the generation of learning curves similar to what is usually done in neural network research)?
2. Which data do you collect during model evaluation and how do you store it?
3. How big is the time-complexity blowup?

I will probably be able to take a closer look on your code changes on Monday.

**Comment by vonHartz** [comment]
1. Exactly! In CCobra there are several learning phases.
- In the evaluation phase learning curves can be generated very literally by evaluating the model on a dataset after each call to the models adapt function,  giving an impression of the learning process of the model.
- In the pretrain or start_participant phase (or within one single call to adapt) the model itself follows some kind of learning strategy. E.g., a model based on some kind of neural network architecture might perform several training epochs on the pre-training data (or the current adaption target). Thus, also tracking these phases might be of interest.

2. I collect a collection of 'model_checkpoints' (which at the moment are just deep copies of the model at at specific time). During the evaluation phase, the Evaluator itself can very easily store them in a list. For the other phases I pass an empty list to the corresponding method (eg pre_train) and let it append it's checkpoint to that list whenever it wants. So, the neural net example can store a checkpoint each epoch (or every 10 or whatever) if it wants to.
All these checkpoints get stored in a single nested dictionary (by phase and subject) and evaluated at the end of the evaluation of the model.
The issue with this approach is, that storing all these checkpoints and only evaluating them once at the end of the model's evaluation really blows up the memory consumption, so evaluating them on the run and only storing the result might be a more clever way of doing things. Changing to this style should be trivial. I will look into that.

3. This really depends on the model and which phases it wants to track. Generating the data for the learning curves for the evaluation phase should be something like Theta(|Models| * |Subjects| * |Items| * (cost(Evaluation) + c(Copy_Model))), generating the very same number of data points. Tracking the learning curves for the internal adaption process with a checkpoint per epoch would make it Theta(|Models| * |Subjects| * |Items| * (cost(Evaluation) + c(Copy_Model) * |Epochs|). I don't know the complexity of preparing and plotting the data by heart (I would have to look into the pandas implementation for that), but it is based on groupby operations, which are probably something like O(n log n) with n being the number of data points.

**Comment by vonHartz** [comment]
I changed it to evaluate on the run now, because storing all the model checkpoints was quite a memory issue.

**Comment by nriesterer** [comment]
We finally found the time to look at your code. Sorry for taking so long.

We see the value in evaluating the change in performance for models. However, we think that the different stages of learning need to be distinguished. First and foremost, CCOBRA is a model benchmarking framework. As such, it is concerned with evaluating the accuracy of models' predictions on test data. What you propose is concerned with evaluating the internal properties of models, i.e., their training performances. These do not matter per se for benchmarking purposes but are important only for the model-engineering perspective.

Your approach has a couple of drawbacks:

1. On a fundamental level, the matter of training performance is only useful for a limited amount of models, namely data-driven ones.
2. On a more technical level, your changes introduce an additional hook for models to interact with CCOBRA by providing the `checkpoints` parameter which is passed to most CCOBRA-related functions.

In our opinion it would be worthwhile to think about ways to directly integrate an evaluation of adaption performance into CCOBRA. Since this can be done without affecting the model interface and is a property of the predictive accuracy a model achieves, it offers additional insight into the performance of model without negatively affecting the complexities of model evaluation and implementation.
